iree/compiler/Dialect/HAL/Transforms/IdentifyConstantPools.cpp

```diff
@@ -0,0 +1,268 @@
+            attr, targetBackend->queryBufferConstraints(context));
+      } else {
+        attr = targetBackend->queryBufferConstraints(context);
+      }
+    }
+    return attr;
+  }
+
+  // Makes a new hal.constant_pool containing the values of the given
+  // variable ops. The variables will be erased and all variable loads will be
+  // replaced with constant loads. Returns the constant pool, if it was created.
+  Optional<ConstantPoolOp> makeConstantPool(
+      StringRef poolName, ArrayRef<IREE::Flow::VariableOp> variableOps,
+      BufferConstraintsAttr bufferConstraints, SymbolTable &moduleSymbolTable,
+      OpBuilder &moduleBuilder) {
+    // Create the pool to be filled with constant values.
+    auto poolOp = OpBuilder(moduleBuilder.getContext())
+                      .create<ConstantPoolOp>(moduleBuilder.getUnknownLoc(),
+                                              poolName, bufferConstraints);
+    moduleSymbolTable.insert(poolOp, moduleBuilder.getInsertionPoint());
+    SymbolTable::setSymbolVisibility(poolOp, SymbolTable::Visibility::Private);
+
+    auto poolBuilder = OpBuilder::atBlockBegin(poolOp.getBody());
+    for (auto variableOp : variableOps) {
+      // Grab the constant value from the variable that we'll be pooling.
+      auto value = variableOp.initial_value()
+                       .getValue()
+                       .dyn_cast_or_null<ElementsAttr>();
+      assert(value && "value precondition not met: must be elements attr");
+
+      // Create the constant in the pool.
+      auto valueOp = poolBuilder.create<ConstantPoolValueOp>(
+          variableOp.getLoc(), variableOp.getName(), value);
+      SymbolTable::setSymbolVisibility(valueOp,
+                                       SymbolTable::Visibility::Nested);
+
+      // Find all variable loads in the module.
+      // May fail if the variable is used in ways not supported by constant
+      // pools.
+      auto loadOps = findVariableLoadOps(variableOp);
+
+      // If the variable is an immutable constant and used in compatible ways
+      // we can turn them into constant loads instead. These will avoid the
+      // additional runtime overhead of variable lifetime tracking and allow
+      // further optimizations at use sites where we know the values come from
+      // constant memory.
+      if (loadOps.hasValue() && !variableOp.is_mutable()) {
+        // Replace all loads of the variable with loads of the constant.
+        for (auto loadOp : loadOps.getValue()) {
+          replaceVariableLoadWithConstantLoad(loadOp, valueOp);
+        }
+        variableOp.erase();
+      } else {
+        // Build an initializer function to populate the variable with the
+        // constant value on startup.
+        changeToVariableInitializerFunc(variableOp, valueOp);
+      }
+    }
+
+    // Remove the pool if it didn't end up with any constants.
+    if (poolOp.getBody()->front().isKnownTerminator()) {
+      poolOp.erase();
+      return None;
```

Comment:
yeah! I just noticed in #3470 executablesLinked is still there but unused :P I'll remove that.
The code golf I did on the working/tested linkExecutables is what let me refactor this to be much better - I feel like there are still some utilities missing (this could be less code). One thing we can do is add additional methods to our ops - so in this case maybe a `poolOp.hasAnyConstants()` etc. Would be good to do a cleanup like that across all passes.